### PR TITLE
Fix build on g++ 13.1 and clang++ 15

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: test/test_gpu_view.cpp [hunk 1]
```diff
@@ -1228,7 +1228,7 @@ TEST_F(NVFuserTest, FusionReshapeIdGraph_CUDA) {
   fusion.addOutput(t13);
 
   // Grab the trivial reduced tensor from t12's reshape.
-  ir_utils::producerTvsOf(tv12)[0];
+  ir_utils::producerTvsOf(tv12);
```
Comment (naoyam) [inline test/test_gpu_view.cpp:1231]: Why is this call? Looks like a no-op.

Comment (zasdfgbnm) [reply test/test_gpu_view.cpp]: Yeah, we should remove the entire line. The error message is:
```C++
/usr/bin/c++ -DUSE_C10D_GLOO -DUSE_C10D_NCCL -DUSE_DISTRIBUTED -DUSE_GTEST -DUSE_RPC -DUSE_TENSORPIPE -I/home/gaoxiang/Fuser2/cmake/../third_party/benchmark/include -I/home/gaoxiang/Fuser2 -I/home/gaoxiang/Fuser2/csrc -I/home/gaoxiang/Fuser2/third_party/flatbuffers/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/googletest/googlemock/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/googletest/googletest/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/flatbuffers/include -isystem /home/gaoxiang/Fuser2/third_party/googletest/googletest/include -isystem /home/gaoxiang/Fuser2/third_party/googletest/googletest -isystem /home/gaoxiang/Fuser2/third_party/googletest/googlemock/include -isystem /home/gaoxiang/Fuser2/third_party/googletest/googlemock -isystem /home/gaoxiang/.local/lib/python3.11/site-packages/torch/include -isystem /home/gaoxiang/.local/lib/python3.11/site-packages/torch/include/torch/csrc/api/include -isystem /usr/local/cuda/include -D_GLIBCXX_USE_CXX11_ABI=0 -O2 -g -DNDEBUG -std=gnu++17 -Wall -Wno-unused-function -D_GLIBCXX_USE_CXX11_ABI=0 -Werror -MD -MT CMakeFiles/nvfuser_tests.dir/test/test_gpu_view.cpp.o -MF CMakeFiles/nvfuser_tests.dir/test/test_gpu_view.cpp.o.d -o CMakeFiles/nvfuser_tests.dir/test/test_gpu_view.cpp.o -c /home/gaoxiang/Fuser2/test/test_gpu_view.cpp
/home/gaoxiang/Fuser2/test/test_gpu_view.cpp: In member function ‘virtual void nvfuser::NVFuserTest_FusionReshapeIdGraph_CUDA_Test::TestBody()’:
/home/gaoxiang/Fuser2/test/test_gpu_view.cpp:1231:34: error: ignoring return value of ‘std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = nvfuser::TensorView*; _Alloc = std::allocator<nvfuser::TensorView*>; reference = nvfuser::TensorView*&; size_type = long unsigned int]’, declared with attribute ‘nodiscard’ [-Werror=unused-result]
 1231 |   ir_utils::producerTvsOf(tv12)[0];
      |                                  ^
In file included from /usr/include/c++/13.1.1/vector:66,
                 from /home/gaoxiang/Fuser2/third_party/googletest/googlemock/include/gmock/gmock-matchers.h:269,
                 from /home/gaoxiang/Fuser2/test/test_gpu_view.cpp:8:
/usr/include/c++/13.1.1/bits/stl_vector.h:1123:7: note: declared here
 1123 |       operator[](size_type __n) _GLIBCXX_NOEXCEPT
      |       ^~~~~~~~
cc1plus: all warnings being treated as errors
```

File: csrc/partition.cpp
Comment (zasdfgbnm) [inline csrc/partition.cpp:1]:
```
/usr/bin/c++ -DTORCH_CUDA_BUILD_MAIN_LIB -DUSE_C10D_GLOO -DUSE_C10D_NCCL -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -Dnvfuser_codegen_EXPORTS -I/home/gaoxiang/Fuser2/cmake/../third_party/benchmark/include -I/home/gaoxiang/Fuser2/third_party/gloo -I/home/gaoxiang/Fuser2/csrc -I/home/gaoxiang/Fuser2/build/include -I/home/gaoxiang/Fuser2/third_party/flatbuffers/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/googletest/googlemock/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/googletest/googletest/include -isystem /home/gaoxiang/Fuser2/cmake/../third_party/flatbuffers/include -isystem /usr/local/cuda/include -isystem /home/gaoxiang/.local/lib/python3.11/site-packages/torch/include -isystem /home/gaoxiang/.local/lib/python3.11/site-packages/torch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=0 -O2 -g -DNDEBUG -std=gnu++17 -fPIC -Wall -Wno-unused-function -DTORCH_CUDA_BUILD_MAIN_LIB -Werror -D_GLIBCXX_USE_CXX11_ABI=0 -MD -MT CMakeFiles/nvfuser_codegen.dir/csrc/partition.cpp.o -MF CMakeFiles/nvfuser_codegen.dir/csrc/partition.cpp.o.d -o CMakeFiles/nvfuser_codegen.dir/csrc/partition.cpp.o -c /home/gaoxiang/Fuser2/csrc/partition.cpp
/home/gaoxiang/Fuser2/csrc/partition.cpp: In function ‘bool nvfuser::{anonymous}::checkInputTensorTypes(const torch::jit::Node*)’:
/home/gaoxiang/Fuser2/csrc/partition.cpp:202:17: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  202 |     const auto& val = node->inputs()[i];
      |                 ^~~
/home/gaoxiang/Fuser2/csrc/partition.cpp:202:39: note: the temporary was destroyed at the end of the full expression ‘node->torch::jit::Node::inputs().c10::ArrayRef<const torch::jit::Value*>::operator[](((long unsigned int)i))’
  202 |     const auto& val = node->inputs()[i];
      |                                       ^
/home/gaoxiang/Fuser2/csrc/partition.cpp: In function ‘bool nvfuser::{anonymous}::checkOutputTensorTypes(const torch::jit::Node*)’:
/home/gaoxiang/Fuser2/csrc/partition.cpp:220:17: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  220 |     const auto& val = node->outputs()[i];
      |                 ^~~
/home/gaoxiang/Fuser2/csrc/partition.cpp:220:40: note: the temporary was destroyed at the end of the full expression ‘node->torch::jit::Node::outputs().c10::ArrayRef<const torch::jit::Value*>::operator[](((long unsigned int)i))’
  220 |     const auto& val = node->outputs()[i];
```

File: csrc/device_lower/pass/scalar_hoist.cpp [hunk 2]
```diff
@@ -458,7 +458,7 @@ class CommonIndexInserter : private kir::ExprMutator {
       // kernel, which can be either int64_t or int. Not very clean,
       // but this seems to be the quickest way to use the value type
       // as we don't have a scalar IR node for the value type.
-      auto dtype = *value->getDataType();
```
Comment (zasdfgbnm) [inline csrc/device_lower/pass/scalar_hoist.cpp:461]:
```C++
    inlined from ‘nvfuser::DataType::DataType(const nvfuser::DataType&)’ at /home/gaoxiang/Fuser2/csrc/type.h:98:8,
    inlined from ‘void nvfuser::{anonymous}::CommonIndexInserter::maybeInsertAllocation(nvfuser::kir::ForLoop*)’ at /home/gaoxiang/Fuser2/csrc/device_lower/pass/scalar_hoist.cpp:462:50:
/usr/include/c++/13.1.1/bits/shared_ptr_base.h:1075:19: error: ‘((const std::__shared_count<__gnu_cxx::_S_atomic>*)((char*)&dtype + offsetof(nvfuser::DataType, nvfuser::DataType::type.std::variant<nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Variant_base<nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Move_assign_base<false, nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Copy_assign_base<false, nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Move_ctor_base<false, nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Copy_ctor_base<false, nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::<unnamed>.std::__detail::__variant::_Variant_storage<false, nvfuser::PrimDataType, nvfuser::ArrayOf, nvfuser::PointerOf>::_M_u)))[1].std::__shared_count<>::_M_pi’ may be used uninitialized [-Werror=maybe-uninitialized]
 1075 |       : _M_pi(__r._M_pi)
      |               ~~~~^~~~~
/home/gaoxiang/Fuser2/csrc/device_lower/pass/scalar_hoist.cpp: In member function ‘void nvfuser::{anonymous}::CommonIndexInserter::maybeInsertAllocation(nvfuser::kir::ForLoop*)’:
/home/gaoxiang/Fuser2/csrc/device_lower/pass/scalar_hoist.cpp:461:12: note: ‘dtype’ declared here
  461 |       auto dtype = *value->getDataType();
```

File: test/test_gpu_view.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply test/test_gpu_view.cpp]: Maybe better, at some point, we should revisit our utilities and add `nodiscard` qualifiers when make sense.
https://en.cppreference.com/w/cpp/language/attributes/nodiscard
We can consider practicing this by adding this qualifier when a new PRs adds a new function.

Comment (zasdfgbnm) [reply test/test_gpu_view.cpp]: removed and the test still pass

File: csrc/partition.cpp
Comment (zasdfgbnm) [reply csrc/partition.cpp:1]: I think this error is like the following

```C++
std::vector<int> one_two() {
  return {1, 2};
}
int i = one_two()[1]; // valid
int& r = one_two()[1]; // valid
```

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**